### PR TITLE
Move Quark SE sensor subsystem to DTS

### PR DESCRIPTION
**Pull request by SavinayDharmappa**
These set of patch add dts support for arc in quark_se_c1000_devboards

### CODE REVIEW

File: dts/arc/quark_se_c1000_ss.dtsi [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#include "mem.h"
+
+/ {
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cpu@0 {
+			device_type = "cpu";
+			compatible = "arc";
```
Comment (galak) [inline dts/arc/quark_se_c1000_ss.dtsi]: take a look at the linux dts for arc, guessing this should be something like "snps,arcXXXXX";

File: dts/arc/quark_se_c1000_ss.dtsi [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+		reg = <0x40000000 DT_FLASH_SIZE>;
+	};
+
+	sram0: memory@a8000400 {
+		device_type = "memory";
+		compatible = "mmio-sram";
+		reg = <0xa8000400 DT_SRAM_SIZE>;
+	};
+
+
```
Comment (galak) [inline dts/arc/quark_se_c1000_ss.dtsi]: remove blank line

File: dts/arc/yaml/intel,qmsi_ss-uart.yaml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      generation: define
+      category: required
+
+  - interrupts:
+      type: array
+      category: required
+      description: required interrupts
+      generation: define
+
+  - pinctrl-\d+:
```
Comment (galak) [inline dts/arc/yaml/intel,qmsi_ss-uart.yaml]: remove clock & pin since you don't use any of that right now

File: dts/arc/quark_se_c1000_ss.dtsi [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	soc {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "simple-bus";
+		ranges;
+
+		uart0: uart@b0002000 {
+			compatible = "intel,qmsi_ss-uart";
```
Comment (galak) [inline dts/arc/quark_se_c1000_ss.dtsi]: Seems like we are using same driver, and not a SS specific one, so any reason this shouldn't just be:

"intel,qmsi-uart"

**Review by galak — CHANGES_REQUESTED** [review]
some changes need to dts, figure out if we should have a unique serial compat for SS vs std qmsi

File: dts/arc/quark_se_c1000_ss.dtsi (hunk 1 above)
Comment (SavinayDharmappa) [reply dts/arc/quark_se_c1000_ss.dtsi]: ok.

File: dts/arc/quark_se_c1000_ss.dtsi (hunk 2 above)
Comment (SavinayDharmappa) [reply dts/arc/quark_se_c1000_ss.dtsi]: ok

File: dts/arc/yaml/intel,qmsi_ss-uart.yaml (hunk 3 above)
Comment (SavinayDharmappa) [reply dts/arc/yaml/intel,qmsi_ss-uart.yaml]: ok will remove it

File: dts/arc/quark_se_c1000_ss.dtsi (hunk 4 above)
Comment (SavinayDharmappa) [reply dts/arc/quark_se_c1000_ss.dtsi]: yes. I will correct it

**Comment by SavinayDharmappa** [comment]
please review this patch

File: dts/arc/mem.h [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#ifndef __DT_BINDING_ST_MEM_H
+#define __DT_BINDING_ST_MEM_H
+
+#define __SIZE_K(x) (x * 1024)
+
+#if defined(CONFIG_SOC_QUARK_SE_C1000_SS)
+#define DT_FLASH_SIZE		__SIZE_K(152)
```
Comment (nashif) [inline dts/arc/mem.h]: can we please skip the mem.h and define this directly in the board dts? I would do the same for x86, this way we will have all information in one place, I know it was done like this for some arm boards but I do not think it is justified here.

**Review by nashif — CHANGES_REQUESTED** [review]

**Review by galak — CHANGES_REQUESTED** [review]
We should get DCCM from device tree as well

**Comment by SavinayDharmappa** [comment]
@nashif @galak  I have incorporated the review comments. Please review the patch

File: dts/arc/quark_se_c1000_ss.dtsi [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+		};
+
+		uart1: uart@b0002400 {
+			compatible = "intel,qmsi-uart";
+			reg = <0xb0002400 0x400>;
+			label = "UART_1";
+
+			status = "disabled";
+		};
+
```
Comment (galak) [inline dts/arc/quark_se_c1000_ss.dtsi]: remove blank line

File: dts/arc/yaml/intel,qmsi_ss-uart.yaml [hunk 7]
```diff
@@ -0,0 +1,30 @@
+---
+title: Intel QMSI Uart
```
Comment (galak) [inline dts/arc/yaml/intel,qmsi_ss-uart.yaml]: move this to dts/bindings/serial

File: dts/arc/yaml/intel,qmsi_ss-uart.yaml [hunk 8]
```diff
@@ -0,0 +1,30 @@
+---
```
Comment (galak) [inline dts/arc/yaml/intel,qmsi_ss-uart.yaml]: either the filename should be intel,qmsi-uart.yaml or the Title, id, compatible constraint all need fixing

File: dts/arc/quark_se_c1000_ss_devboard.fixup [hunk 9]
```diff
@@ -0,0 +1,17 @@
+#define CONFIG_UART_QMSI_0_BAUDRATE	INTEL_QMSI_UART_B0002000_CURRENT_SPEED
+#define CONFIG_UART_QMSI_0_NAME		INTEL_QMSI_UART_B0002000_LABEL
+
+#define CONFIG_UART_QMSI_1_BAUDRATE	INTEL_QMSI_UART_B0002400_CURRENT_SPEED
+#define CONFIG_UART_QMSI_1_NAME		INTEL_QMSI_UART_B0002400_LABEL
+
+#define SRAM_START			CONFIG_SRAM_BASE_ADDRESS
+
```
Comment (galak) [inline dts/arc/quark_se_c1000_ss_devboard.fixup]: remove blank line

File: dts/arc/quark_se_c1000_ss_devboard.fixup [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+#define SRAM_SIZE			CONFIG_SRAM_SIZE
+
+#define FLASH_START			CONFIG_FLASH_BASE_ADDRESS
+
```
Comment (galak) [inline dts/arc/quark_se_c1000_ss_devboard.fixup]: remove blank line

File: dts/arc/quark_se_c1000_ss_devboard.fixup [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+#define FLASH_SIZE      		CONFIG_FLASH_SIZE
+
+#define DCCM_START			CONFIG_DCCM_BASE_ADDRESS
+
```
Comment (galak) [inline dts/arc/quark_se_c1000_ss_devboard.fixup]: remove blank line

File: dts/arc/quark_se_c1000_ss_devboard.dts [hunk 12]
```diff
@@ -0,0 +1,35 @@
+/dts-v1/;
+
+#define __SIZE_K(x) (x * 1024)
+
+#if defined(CONFIG_SOC_QUARK_SE_C1000_SS)
+#define DT_FLASH_SIZE		__SIZE_K(152)
+#define DT_SRAM_SIZE		__SIZE_K(24)
+#define DT_DCCM_SIZE		__SIZE_K(8)
+#else
+#error "Flash and RAM sizes not defined for this chip"
```
Comment (galak) [inline dts/arc/quark_se_c1000_ss_devboard.dts]: why is this here, if you want the #defines, than remove the #if, #else, #error as it doesn't serve any purpose

File: scripts/dts/extract_dts_includes.py [hunk 13]
```diff
@@ -798,6 +798,10 @@ def main():
         extract_reg_prop(chosen['zephyr,sram'], None,
                          defs, "CONFIG_SRAM", 1024, None)
 
+    if 'zephyr,dccm' in chosen:
```
Comment (galak) [inline scripts/dts/extract_dts_includes.py]: no, dccm should be done via a yaml

**Review by galak — CHANGES_REQUESTED** [review]
Number of cleanup comments, and we need to change how we generate dccm info.

Also, is there a reason ICCM is missing?

File: dts/arc/quark_se_c1000_ss.dtsi (hunk 6 above)
Comment (SavinayDharmappa) [reply dts/arc/quark_se_c1000_ss.dtsi]: k

File: dts/arc/quark_se_c1000_ss_devboard.dts (hunk 12 above)
Comment (SavinayDharmappa) [reply dts/arc/quark_se_c1000_ss_devboard.dts]: will change it

File: dts/arc/yaml/intel,qmsi_ss-uart.yaml (hunk 8 above)
Comment (SavinayDharmappa) [reply dts/arc/yaml/intel,qmsi_ss-uart.yaml]: k

File: dts/arc/yaml/intel,qmsi_ss-uart.yaml (hunk 7 above)
Comment (SavinayDharmappa) [reply dts/arc/yaml/intel,qmsi_ss-uart.yaml]: will move it

File: scripts/dts/extract_dts_includes.py (hunk 13 above)
Comment (SavinayDharmappa) [reply scripts/dts/extract_dts_includes.py]: ok

**Comment by SavinayDharmappa** [comment]
@galak  ICCM is not part of arc in quark_sec_c1000 .

**Comment by galak** [comment]
> @galak ICCM is not part of arc in quark_sec_c1000 .

Gotcha, if we introduce a yaml for DCCM, we should probably do one for ICCM to match, even if quark_sec_c1000 doesn't support it.

**Comment by SavinayDharmappa** [comment]
@galak incorporated the changes please review it.

File: dts/arc/quark_se_c1000_ss_devboard.dts [hunk 14]
```diff
@@ -0,0 +1,30 @@
+/dts-v1/;
+
+#define __SIZE_K(x) (x * 1024)
```
Comment (galak) [inline dts/arc/quark_se_c1000_ss_devboard.dts]: This should be in the soc .dtsi not here.

File: dts/arc/quark_se_c1000_ss.dtsi [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cpu@0 {
+			device_type = "cpu";
+			compatible = "snps,arcem4";
+			reg = <1>;
+		};
+	};
+
```
Comment (galak) [inline dts/arc/quark_se_c1000_ss.dtsi]: We really should have the ARC interrupt controller

File: dts/arc/quark_se_c1000_ss.dtsi [hunk 16]
```diff
@@ -0,0 +1,53 @@
+#include "skeleton.dtsi"
+
+/ {
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cpu@0 {
+			device_type = "cpu";
+			compatible = "snps,arcem4";
```
Comment (galak) [inline dts/arc/quark_se_c1000_ss.dtsi]: is the cpu really em4?

File: dts/arc/quark_se_c1000_ss.dtsi [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+
+	sram0: memory@a8000400 {
+		device_type = "memory";
+		compatible = "mmio-sram";
+		reg = <0xa8000400 DT_SRAM_SIZE>;
+	};
+
+	dccm0: dccm@80000000 {
+		device_type = "memory";
+		compatible = "intel,qmsi_ss-dccm";
```
Comment (galak) [inline dts/arc/quark_se_c1000_ss.dtsi]: are these really properties specific to intel,qmsi? seem like the are arc related?

File: dts/arc/quark_se_c1000_ss.dtsi [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "simple-bus";
+		ranges;
+
+		uart0: uart@b0002000 {
+			compatible = "intel,qmsi-uart";
+			reg = <0xb0002000 0x400>;
+			label = "UART_0";
+
```
Comment (galak) [inline dts/arc/quark_se_c1000_ss.dtsi]: we should have interrupts for the uart's

**Review by galak — CHANGES_REQUESTED** [review]
Need to fixup interrupt support in dts for arch

File: dts/arc/quark_se_c1000_ss.dtsi (hunk 16 above)
Comment (SavinayDharmappa) [reply dts/arc/quark_se_c1000_ss.dtsi]: yes. That is what mentioned in data sheet of quark_se_c1000

Comment (galak) [reply dts/arc/quark_se_c1000_ss.dtsi]: Can you point me at the data sheet for the quark_se_c1000?

Comment (SavinayDharmappa) [reply dts/arc/quark_se_c1000_ss.dtsi]: https://www.intel.com/content/dam/www/public/us/en/documents/datasheets/quark-c1000-datasheet.pdf.

In first page only it is mentioned "ARC EM4 DSP with floating point 
extensions"

File: dts/arc/quark_se_c1000_ss.dtsi (hunk 15 above)
Comment (SavinayDharmappa) [reply dts/arc/quark_se_c1000_ss.dtsi]: ok

File: dts/arc/quark_se_c1000_ss.dtsi (hunk 17 above)
Comment (SavinayDharmappa) [reply dts/arc/quark_se_c1000_ss.dtsi]: i will correct it

**Comment by SavinayDharmappa** [comment]
@galak  i have added interrupt node to arc. Could you please review it.

File: dts/arc/quark_se_c1000_ss.dtsi [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+		#size-cells = <0>;
+
+		cpu@0 {
+			device_type = "cpu";
+			compatible = "snps,arcem4";
+			reg = <1>;
+		};
+
+		core_intc: arcem4-intc@0 {
+			compatible = "snps,arcem4-intc";
```
Comment (galak) [inline dts/arc/quark_se_c1000_ss.dtsi]: lets rename this to:
compatible = "snps,arcv2-intc";

File: dts/arc/quark_se_c1000_ss_devboard.fixup [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+#define SRAM_START			CONFIG_SRAM_BASE_ADDRESS
+#define SRAM_SIZE			CONFIG_SRAM_SIZE
+
+#define FLASH_START			CONFIG_FLASH_BASE_ADDRESS
+#define FLASH_SIZE      		CONFIG_FLASH_SIZE
+
+#define CONFIG_DCCM_BASE_ADDRESS       ARC_DCCM_80000000_BASE_ADDRESS
+#define CONFIG_DCCM_SIZE               ARC_DCCM_80000000_SIZE
+
+#define CONFIG_ICCM_BASE_ADDRESS       ARC_ICCM_FFFFFFFF_BASE_ADDRESS
```
Comment (galak) [inline dts/arc/quark_se_c1000_ss_devboard.fixup]: since we don't have ICCM why add these?

File: dts/bindings/interrupt-controller/snps,arcem4-intc.yaml [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+
+description: >
+    This binding describes the ARC EM4 IRQ controller
+
+properties:
+  - compatible:
+      category: required
+      type: string
+      description: compatible strings
+      constraint: "snps,arcem4-intc"
```
Comment (galak) [inline dts/bindings/interrupt-controller/snps,arcem4-intc.yaml]: see comment above about change of compat

**Review by galak — CHANGES_REQUESTED** [review]
some minor fixup/cleanups.

File: dts/arc/quark_se_c1000_ss.dtsi (hunk 19 above)
Comment (SavinayDharmappa) [reply dts/arc/quark_se_c1000_ss.dtsi]: ok

**Comment by SavinayDharmappa** [comment]
@galak  incorporated all the review comments.

**Comment by SavinayDharmappa** [comment]
@galak can this be merged ?

**Review by galak — CHANGES_REQUESTED** [review]
Can you update for cmake change, also add license header to the yaml's (see dts/bindings/arm/nxp,kinetis-sim.yaml)

File: dts/arc/mem.h (hunk 5 above)
Comment (galak) [reply dts/arc/mem.h]: Its partially a question if the memory & flash sizes are a function of the board or SoC.

Comment (SavinayDharmappa) [reply dts/arc/mem.h]: I think reason is quark_se_c1000 soc based borads arduino_101 and tinytile has different flash size so it make sense to move it to the board. so it is a function of board

**Comment by SavinayDharmappa** [comment]
@galak  incorporated cmake changes and license headers.

**Comment by SavinayDharmappa** [comment]
@galak can this be merged ?

**Review by nashif — APPROVED** [review]

**Review by galak — CHANGES_REQUESTED** [review]
Do we still need the Makefile?

also duplicate between board/arc/quark_se_c1000_ss_devboard/dts.fixup vs dts/arc/quark_se_c1000_ss_devboard.fixup

**Comment by SavinayDharmappa** [comment]
@galak  i have removed .dts, .fixup and makefile from dts/ folder. flashing was failing for arc on quark_devboard it is also fixed.

**Review by galak — APPROVED** [review]
